src/WebJobs.Script/Host/ProxyFunctionProvider.cs

```diff
@@ -0,0 +1,195 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Collections.ObjectModel;
+using System.IO;
+using System.Linq;
+using System.Reactive.Linq;
+using System.Text.RegularExpressions;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Azure.AppService.Proxy.Client;
+using Microsoft.Azure.WebJobs.Logging;
+using Microsoft.Azure.WebJobs.Script.Description;
+using Microsoft.Azure.WebJobs.Script.Eventing;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.Azure.WebJobs.Script
+{
+    public class ProxyFunctionProvider : IFunctionProvider, IDisposable
+    {
+        private static readonly Regex ProxyNameValidationRegex = new Regex(@"[^a-zA-Z0-9_-]", RegexOptions.IgnoreCase);
+        private readonly ReaderWriterLockSlim _metadataLock = new ReaderWriterLockSlim();
+        private readonly IOptions<ScriptJobHostOptions> _scriptOptions;
+        private readonly IEnvironment _environment;
+        private readonly ILogger _logger;
+        private readonly IDisposable _fileChangeSubscription;
+        private Dictionary<string, ICollection<string>> _functionErrors = new Dictionary<string, ICollection<string>>();
+        private Lazy<ImmutableArray<FunctionMetadata>> _metadata;
+        private bool _disposed = false;
+
+        public ProxyFunctionProvider(IOptions<ScriptJobHostOptions> scriptOptions, IEnvironment environment, IScriptEventManager eventManager, ILoggerFactory loggerFactory)
+        {
+            _scriptOptions = scriptOptions;
+            _environment = environment;
+            _logger = loggerFactory.CreateLogger(LogCategories.Startup);
+            _metadata = new Lazy<ImmutableArray<FunctionMetadata>>(LoadFunctionMetadata);
+
+            _fileChangeSubscription = eventManager.OfType<FileEvent>()
+                       .Where(f => string.Equals(f.Source, EventSources.ScriptFiles, StringComparison.Ordinal) &&
+                       string.Equals(Path.GetFileName(f.FileChangeArguments.Name), ScriptConstants.ProxyMetadataFileName, StringComparison.OrdinalIgnoreCase))
+                       .Subscribe(e => HandleProxyFileChange());
+        }
+
+        public ImmutableDictionary<string, ImmutableArray<string>> FunctionErrors => _functionErrors.ToImmutableDictionary(kvp => kvp.Key, kvp => kvp.Value.ToImmutableArray());
```

Comment:
Instead of building this each time, can we just store it as immutable? It looks like `_functionErrors` is written to only once.

Reply:
fixed in c34f9fe5078d77d670890be613c023113a3862eb